internal/releasesources/release_sources.go

```diff
@@ -0,0 +1,413 @@
+	if count == 1 {
+		// There is a release to merge with, add the CPU architecture to the array if needed
+		var release common.OpenshiftVersion
+		err := db.Model(common.OpenshiftVersion{}).
+			Where("display_name = ? AND channel = ?", displayName, channel).
+			First(&release).Error
+		if err != nil {
+			return err
+		}
+
+		if !contains(release.CpuArchitectures, cpuArchitecture) {
+			release.CpuArchitectures = append(release.CpuArchitectures, cpuArchitecture)
+			return db.Save(&release).Error
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("more than one match found with openshift_version = %s, channel = %s", displayName, channel)
+}
+
+func contains(slice []string, str string) bool {
+	for _, v := range slice {
+		if v == str {
+			return true
+		}
+	}
+	return false
+}
+
+func syncReleaseImages(db *gorm.DB, releaseSources models.ReleaseSources, client OpenShiftReleasesAPIClient) error {
+	// OCP releases API needs amd64, arm64 instead of x86_64, aarch64 respectively
+	cpuArchMapToAPIArch := map[string]string{
+		"x86_64":  "amd64",
+		"aarch64": "arm64",
+	}
+
+	// switch back
+	cpuArchMapFromAPIArch := map[string]string{
+		"amd64": "x86_64",
+		"arm64": "aarch64",
+	}
+
+	var switchedArch bool = false
+
+	for _, releaseSource := range releaseSources {
+		openshiftVersion := *releaseSource.OpenshiftVersion
+		for _, upgradeChannel := range releaseSource.UpgradeChannels {
+			cpuArchitecture := *upgradeChannel.CPUArchitecture
+			for _, channel := range upgradeChannel.Channels {
+				if arch, ok := cpuArchMapToAPIArch[cpuArchitecture]; ok {
+					switchedArch = true
+					cpuArchitecture = arch
+				}
+
+				graph, err := client.getReleases(channel, openshiftVersion, cpuArchitecture)
+				if err != nil {
+					return err
+				}
+
+				if switchedArch {
+					if arch, ok := cpuArchMapFromAPIArch[cpuArchitecture]; ok {
+						cpuArchitecture = arch
+					}
+				}
+
+				for _, node := range graph.Nodes {
+					if strings.HasPrefix(node.Version, openshiftVersion) {
+						// Got a match for OpenShiftReleaseImage
+						err = createOrUpdateRelease(db, channel, node.Version, cpuArchitecture)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func deletePreReleaseVersions(db *gorm.DB, openshiftMajorMinorVersion string) error {
+	var (
+		releasesToDelete []common.OpenshiftVersion
+		releases         []common.OpenshiftVersion
+	)
+	err := db.Model(&common.OpenshiftVersion{}).
+		Find(&releases).Error
+	if err != nil {
+		return err
+	}
+
+	for _, release := range releases {
+		majorMinorVersion, err := common.GetMajorMinorVersion(*release.OpenshiftVersion.DisplayName)
+		if err != nil {
+			return err
+		}
+
+		if *majorMinorVersion == openshiftMajorMinorVersion && release.Channel == "candidate" {
+			releasesToDelete = append(releasesToDelete, release)
+		}
+	}
+	// Nothing to delete
+	if len(releasesToDelete) == 0 {
+		return nil
+	}
+
+	return db.Where("TRUE").Delete(&releasesToDelete).Error
```

Comment:
Why where TRUE?

Reply:
I think `Delete` must have Where as a prerequisite, not sure though I will check